pkg/workload/workload_generator/workload.go

```diff
@@ -309,78 +368,106 @@ func generateBatch(
 	if err != nil {
 		return workload.QueryLoad{}, err
 	}
+	db.SetMaxOpenConns(w.connFlags.Concurrency + 1)
+	db.SetMaxIdleConns(w.connFlags.Concurrency + 1)
 
-	ql := workload.QueryLoad{
-		WorkerFns: make([]func(context.Context) error, 8),
-	}
-	initialRowCount := int64(10)
-	for i := range ql.WorkerFns {
-		workerID := i
-		rng := rand.New(rand.NewPCG(uint64(workerID), 0))
-		ql.WorkerFns[i] = func(ctx context.Context) error {
-			// 50% reads, 50% updates
-			if rng.IntN(2) == 0 {
-				// Read operation
-				id := rng.Int64N(initialRowCount)
-				var field1, field2, field3 string
-				if err := readStmt.QueryRowContext(ctx, id).Scan(&field1, &field2, &field3); err != nil {
-					return err
-				}
-			} else {
-				// Update operation
-				id := rng.Int64N(initialRowCount)
-				field1 := randString(rng, fieldLength)
-				field2 := randString(rng, fieldLength)
-				field3 := randString(rng, fieldLength)
-				if _, err := updateStmt.ExecContext(ctx, field1, field2, field3, id); err != nil {
-					return err
-				}
-			}
-			return nil
-		}
+	// The generators and column old data caches are initialised.
+	if err := w.initGenerators(db); err != nil {
+		return workload.QueryLoad{}, err
 	}
 
-	// Close statements and database when the workload finishes
-	ql.Close = func(context.Context) error {
-		err := readStmt.Close()
-		if err != nil {
-			return err
+	ql := workload.QueryLoad{}
+	for i := 0; i < w.connFlags.Concurrency; i++ {
+		worker := &txnWorker{
+			db:                db,
+			readTransactions:  readTransactions,
+			writeTransactions: writeTransactions,
+			rng:               rand.New(rand.NewPCG(uint64(time.Now().UnixNano()), uint64(i))),
+			hists:             reg.GetHandle(),
+			w:                 w, // reference to the workloadGeneratorStruct for generators
 		}
-		err = updateStmt.Close()
-		if err != nil {
-			return err
-		}
-		return sqlDatabase.Close()
+		ql.WorkerFns = append(ql.WorkerFns, worker.run)
 	}
-
+	ql.Close = func(context.Context) error { return db.Close() }
 	return ql, nil
 }
 
-func randString(rng *rand.Rand, length int) []byte {
-	const letters = "abcdefghijklmnopqrstuvwxyz"
-	result := make([]byte, length)
-	for i := range result {
-		result[i] = letters[rng.IntN(len(letters))]
+// run executes a random transaction from the list of transactions.
+func (t *txnWorker) run(ctx context.Context) error {
+	txn := t.chooseTransaction()
+	// Reference to the workloadGeneratorStruct for generators.
+	w := t.w
+	// Time for the metrics is started.
+	start := timeutil.Now()
+
+	// Each transaction gets its own debug slice.
+	type debugEntry struct {
+		Query string
+		Args  []interface{}
 	}
-	return result
+	err := crdb.ExecuteTx(ctx, t.db, nil, func(tx *gosql.Tx) error {
+		// Inserted maintains a map of column names to the values that were inserted in this transaction.
+		inserted := make(map[string][]interface{})
+		queryCtx, cancel := context.WithTimeout(ctx, 5*time.Minute)
```

Comment:
Please make 5 minutes a constant or better an input flag. That way it can be configured, if we know that there are queries that may take longer.